### PR TITLE
sensor: add TI FDC2X1X sensor driver and sample application 

### PR DESCRIPTION
**Pull request by skullbox305**
This patch adds support for the Texas Instruments FDC2X1X Capacitance-to-Digital Converter for Proximity and Level Sensing Applications. 
Along with the driver, a sample sensor application is provided which tests the device driver.

Signed-off-by: Igor Knippenberg <igor.knippenberg@gmail.com>

### CODE REVIEW

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 1]
```diff
@@ -0,0 +1,1010 @@
...
+	struct fdc2x1x_data *fdc2x1x_data = dev->data;
+	const struct fdc2x1x_config *cfg = dev->config;
+
+	if (reg & FDC2X1X_READ) {
+		return i2c_burst_read(fdc2x1x_data->bus,
+				      cfg->i2c_addr,
+				      FDC2X1X_TO_I2C_REG(reg),
+				      (uint8_t *) data, length);
+	} else {
+		if (length <= 0) {
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:42]: What's going on here? The `length` argument is a `size_t`, which is unsigned, so this doesn't look right.

Should this be something like `if (length != 2)` instead? The `memcpy(buf + 1, data, sizeof(uint16_t));` line below makes it seem like exactly two bytes of `data` are being used in this case.

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 2]
```diff
@@ -0,0 +1,1010 @@
...
+ * Read/Write from device.
+ * @param dev - The device structure.
+ * @param reg - The register address. Use FDC2X1X_REG_READ(x) or
+ *				FDC2X1X_REG_WRITE(x).
+ * @param data - The register data.
+ * @param length - Number of bytes being read
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int fdc2x1x_bus_access(const struct device *dev, uint8_t reg,
+			      void *data, size_t length)
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:31]: Why not just declare this `uint8_t *data` and avoid casting it to `uint8_t*` below? The only two callers are allocating uint8_t arrays for this argument.

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 3]
```diff
@@ -0,0 +1,1010 @@
...
+	return 0;
+}
+
+/**
+ * Enable/disable Shutdown Mode
+ * @param dev - The device structure.
+ * @param enable - True = enable shutdown, false = disable shutdown
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int fdc2x1x_set_shutdown(const struct device *dev, bool enable)
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:431]: This should be `static`.

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 4]
```diff
@@ -0,0 +1,1010 @@
...
+{
+	struct fdc2x1x_data *data = dev->data;
+	uint32_t new_state;
+	int ret = 0;
+
+	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
+		new_state = *(uint32_t *)context;
+
+		if (new_state != data->pm_state) {
+
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:517]: Nit: unnecessary whitespace.

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 5]
```diff
@@ -0,0 +1,1010 @@
...
+	return ret;
+}
+#endif
+
+/**
+ * Set attributes for the device.
+ * @param dev - The device structure.
+ * @param chan - The sensor channel type.
+ * @param attr - The sensor atrribute.
+ * @param value - The sensor atrribute value.
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:546]: Nit: "attribute" typos

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 6]
```diff
@@ -0,0 +1,1010 @@
...
+
+	/* Channel specific settings */
+	for (ch = 0; ch < cfg->num_channels; ch++) {
+		ret = fdc2x1x_set_fin_sel(dev, ch, cfg->ch_cfg[ch].fin_sel);
+		if (ret) {
+			return ret;
+		}
+
+		ret =
+			fdc2x1x_set_fref_divider(dev, ch,
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:716]: Nit: this looks like an unnecessary extra newline.

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 7]
```diff
@@ -0,0 +1,1010 @@
...
+	data->gpio = device_get_binding(cfg->sd_name);
+	if (!data->gpio) {
+		LOG_ERR("Failed to get pointer to %s device!", cfg->sd_name);
+		return -EINVAL;
+	}
+
+	gpio_pin_configure(data->gpio, cfg->sd_pin,
+			   GPIO_OUTPUT_INACTIVE | cfg->sd_flags);
+
+	gpio_pin_set(data->gpio, cfg->sd_pin, 0);
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:864]: This seems redundant with the GPIO_OUTPUT_INACTIVE flag above, no?

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 8]
```diff
@@ -0,0 +1,1010 @@
...
+	const struct fdc2x1x_config *cfg = dev->config;
+	uint8_t ch_supported;
+
+	if (cfg->fdc2x14) {
+		ch_supported = 4;
+	} else {
+		ch_supported = 2;
+	}
+
+	if ((cfg->num_channels <= 0)) {
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:886]: Nit: extraneous parentheses

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 9]
```diff
@@ -0,0 +1,1010 @@
...
+		}
+	}
+
+	data->bus = device_get_binding(cfg->i2c_name);
+	if (data->bus  == NULL) {
+		LOG_ERR("Failed to get pointer to %s device!", cfg->i2c_name);
+		return -EINVAL;
+	}
+
+	k_sleep(K_MSEC(100));
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:906]: Why? This is going to block the init sequence for 100 msec.

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 10]
```diff
@@ -0,0 +1,1010 @@
...
+
+#define FDC2X1X_SD_PROPS(n)			     \
+	.sd_name = DT_INST_GPIO_LABEL(n, sd_gpios),  \
+	.sd_pin = DT_INST_GPIO_PIN(n, sd_gpios),     \
+	.sd_flags = DT_INST_GPIO_FLAGS(n, sd_gpios), \
+
+#define FDC2X1X_SD(n)					\
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(n, sd_gpios),	\
+		    (FDC2X1X_SD_PROPS(n)),		\
+		    ())
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:945]: Feel free to leave this as-is, but this could be written more simply as:

```
	IF_ENABLED(DT_INST_NODE_HAS_PROP(n, sd_gpios),	\
		    (FDC2X1X_SD_PROPS(n)))
```

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 11]
```diff
@@ -0,0 +1,1010 @@
...
+
+#define FDC2X1X_INTB_PROPS(n)				 \
+	.intb_name = DT_INST_GPIO_LABEL(n, intb_gpios),	 \
+	.intb_pin = DT_INST_GPIO_PIN(n, intb_gpios),	 \
+	.intb_flags = DT_INST_GPIO_FLAGS(n, intb_gpios), \
+
+#define FDC2X1X_INTB(n)			     \
+	COND_CODE_1(CONFIG_FDC2X1X_TRIGGER,  \
+		    (FDC2X1X_INTB_PROPS(n)), \
+		    ())
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:955]: Same here. Feel free to leave as-is, but this could use IF_ENABLED also.

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 12]
```diff
@@ -0,0 +1,1010 @@
...
+
+#define FDC2X1X_CH_CFG_INIT(m)				  \
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:957]: I think `ch` would be a better variable name than `m`. Here, `m` after `n` suggests that `m` is also an instance number, but it's actually a node identifier.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+        The INTB pin defaults to active low when produced by the sensor.
+        The property value should ensure the flags properly describe
+        the signal that is presented to the driver.
+
+    fdc2x14:
+      type: boolean
+      description: |
+        Set to identify the sensor as FDC2114 or FDC2214 (4-channel version)
+
+    autoscan:
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:32]: Couldn't this just be boolean?

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+        active-channel.
+        enabled = Auto-Scan conversions as selected by rr-sequence.
+      enum:
+        - "disabled"
+        - "enabled"
+
+    rr-sequence:
+      type: int
+      required: false
+      default: 0
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:48]: Defaults in bindings must be justified; please explain why its value is set in the description.

If this value corresponds to the power-on-reset behavior of the device and the property can be changed independently of other properties, that's OK. If this is an individual preference made to work with the sample application or something like that, the property should not have the default value (and may need to be made `required` instead if there's no sane default corresponding to PoR behavior or something like that).

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+        2 = Ch0, Ch1, Ch2, Ch3 (FDC2114, FDC2214 only)
+      enum:
+        - 0
+        - 1
+        - 2
+
+    active-channel:
+      type: int
+      required: false
+      default: 0
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:66]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+      enum:
+        - 0
+        - 1
+        - 2
+        - 3
+
+    deglitch:
+      type: int
+      required: false
+      default: 5
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:83]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+      enum:
+        - 1
+        - 4
+        - 5
+        - 7
+
+    sensor-activate-sel:
+      type: string
+      required: false
+      default: "low-power"
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:101]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+        programmed by chx_idrive during sensor activation to minimize
+        power consumption.
+      enum:
+        - "full-current"
+        - "low-power"
+
+    ref-clk-src:
+      type: string
+      required: false
+      default: "internal"
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:118]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+        internal = Use Internal oscillator as reference frequency
+        external = Reference frequency is provided from CLKIN pin.
+      enum:
+        - "internal"
+        - "external"
+
+    current-drive:
+      type: string
+      required: false
+      default: "normal"
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:131]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+
+        This mode is not supported if autoscan = "disabled" (multichannel mode)
+      enum:
+        - "normal"
+        - "high"
+
+    output-gain:
+      type: int
+      required: false
+      default: 0
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:148]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+child-binding:
+  description: |
+    Settings for each channel 0-1 (FDC2112 and FDC2212) or
+    0-3 (FDC2114 and FDC2214)
+
+  properties:
+      rcount:
+        type: int
+        required: false
+        default: 7499
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:171]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+        description: |
+          Channel X Reference Count Conversion Interval Time.
+          Valid range: 256 - 65535
+
+      offset:
+        type: int
+        required: false
+        default: 0
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:179]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+        description: |
+          Channel X Conversion Offset (FDC2112 and FDC2212 only).
+          Valid range: 0 - 65535
+
+      settlecount:
+        type: int
+        required: false
+        default: 48
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:187]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+          Channel X Conversion Settling.
+
+          The FDC will use this settling time to allow the LC sensor to
+          stabilize before initiation of a conversion on Channel X.
+          Valid range: 0 - 65535
+
+      fref-divider:
+        type: int
+        required: false
+        default: 1
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:198]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+          Channel X Reference Divider.
+
+          Sets the divider for Channel 3 reference.
+          Use this to scale the maximum conversion frequency.
+          Valid range: 1 - 1023
+
+      idrive:
+        type: int
+        required: false
+        default: 10
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:209]: Same comments about default made above apply here also.

File: dts/bindings/sensor/ti,fdc2x1x.yaml [hunk 26]
```diff
@@ -0,0 +1,235 @@
...
+          Channel X Sensor drive current.
+
+          This field defines the Drive Current used during the settling +
+          conversion time of Channel 3 sensor clock.
+          Valid range: 0 - 31
+
+      fin-sel:
+        type: int
+        required: false
+        default: 2
```
Comment (mbolivar-nordic) [inline dts/bindings/sensor/ti,fdc2x1x.yaml:220]: Same comments about default made above apply here also.

File: include/drivers/sensor.h [hunk 27]
```diff
@@ -174,6 +174,15 @@ enum sensor_channel {
...
 
+	/** Channel 0, raw data */
+	SENSOR_CHAN_RAW_CH0,
+	/** Channel 1, raw data */
+	SENSOR_CHAN_RAW_CH1,
+	/** Channel 2, raw data */
+	SENSOR_CHAN_RAW_CH2,
+	/** Channel 3, raw data */
+	SENSOR_CHAN_RAW_CH3,
+
```
Comment (mbolivar-nordic) [inline include/drivers/sensor.h:185]: I'm not sure this makes sense to add here. @MaureenHelm ?

File: samples/sensor/fdc2x1x/README.rst [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+
+For detailed description refer to the `FDC2X1X datasheet`_
+at pages 4-5.
+
+
+Building and Running
+********************
+
+This sample outputs sensor data to the console and can be read by any serial
+console program. It should work with any platform featuring a I2C interface.
```
Comment (mbolivar-nordic) [inline samples/sensor/fdc2x1x/README.rst:35]: You should add a note that different platforms will require custom devicetree overlays.

File: samples/sensor/fdc2x1x/boards/nrf9160dk_nrf9160.overlay [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+&i2c2 {
+	clock-frequency = <I2C_BITRATE_FAST>;
+	fdc2x1x@2A                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                {
+		compatible = "ti,fdc2x1x";
+		reg = <0x2A>;
+		label = "FDC2X1X";
+		sd-gpios = <&gpio0 16 GPIO_ACTIVE_HIGH>;
+		intb-gpios = <&gpio0 20 GPIO_ACTIVE_LOW>;
+
+		// using the default channel settings
```
Comment (mbolivar-nordic) [inline samples/sensor/fdc2x1x/boards/nrf9160dk_nrf9160.overlay:16]: Judging by this comment, I think the binding should probably not have default settings, but instead the properties should be made `required` and this sample can provide some sensible defaults for evaluation here.

File: samples/sensor/fdc2x1x/src/main.c [hunk 30]
```diff
@@ -0,0 +1,134 @@
...
+	};
+
+	uint8_t num_ch = ARRAY_SIZE(ch_buf);
+	int i;
+
+	const struct device *dev =
+		device_get_binding(DT_LABEL(DT_INST(0, ti_fdc2x1x)));
+
+	if (dev == NULL) {
+		printk("Could not get %s device \n",
```
Comment (mbolivar-nordic) [inline samples/sensor/fdc2x1x/src/main.c:76]: Extra whitespace here.

File: samples/sensor/fdc2x1x/src/main.c [hunk 31]
```diff
@@ -0,0 +1,134 @@
...
+	} else {
+		printk("Success\n");
+	}
+}
+#endif
+
+void main(void)
+{
+	struct sensor_value ch_buf[] = {
+		DT_FOREACH_CHILD(DT_INST(0, ti_fdc2x1x), CH_BUF_INIT)
```
Comment (mbolivar-nordic) [inline samples/sensor/fdc2x1x/src/main.c:66]: @galak @pabigot I'm curious about what you think about the continued use of DT_INST() in sample applications.

I'm not really in favor of it because it means that users have to understand instance numbers, which are complicated and basically just for driver authors. Getting instance numbers to work in a sane way outside of a driver breaks down fast when there are multiple devices of the same compatible.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Thanks for your contribution!

**Comment by mbolivar-nordic** [comment]
One more thing: the driver must at least be build-tested by modifying tests/drivers/build_all.

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 1 above)
Comment (skullbox305) [reply drivers/sensor/fdc2x1x/fdc2x1x.c]: your are right, my bad

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 2 above)
Comment (skullbox305) [reply drivers/sensor/fdc2x1x/fdc2x1x.c]: declared as `uint8_t *data` now

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 7 above)
Comment (skullbox305) [reply drivers/sensor/fdc2x1x/fdc2x1x.c]: yes this call is unecessary. I removed it

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 9 above)
Comment (skullbox305) [reply drivers/sensor/fdc2x1x/fdc2x1x.c]: Forgot to remove it after testing. Removed it now

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 10 above)
Comment (skullbox305) [reply drivers/sensor/fdc2x1x/fdc2x1x.c]: I took your suggestions. It is definitely easier to read

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 12 above)
Comment (skullbox305) [reply drivers/sensor/fdc2x1x/fdc2x1x.c]: changed the variable name to `ch`

File: dts/bindings/sensor/ti,fdc2x1x.yaml (hunk 13 above)
Comment (skullbox305) [reply dts/bindings/sensor/ti,fdc2x1x.yaml:32]: Yes a boolean makes more sense. I changed that

File: dts/bindings/sensor/ti,fdc2x1x.yaml (hunk 14 above)
Comment (skullbox305) [reply dts/bindings/sensor/ti,fdc2x1x.yaml:46]: Most of those values are power-on-reset defaults. The ones which are not I changed to required. I also added an explanation in each description for this and your comments below

File: dts/bindings/sensor/ti,fdc2x1x.yaml (hunk 22 above)
Comment (skullbox305) [reply dts/bindings/sensor/ti,fdc2x1x.yaml:190]: This value applies only to the 12 Bit version. So I dont think it makes sense to make `required`. I leave the default power-on-reset value as `default`

File: include/drivers/sensor.h (hunk 27 above)
Comment (skullbox305) [reply include/drivers/sensor.h:185]: I was also not happy about that, but I don't think the current sensor API is able to read multiple channels of the same type from a single device? Would be good to be able to pass something like a channel number to the channel get function e.g. `sensor_channel_get(dev, SENSOR_CHAN_RAW, ch_number, buffer);`. Is there a plan to add a functionality like that?

File: samples/sensor/fdc2x1x/README.rst (hunk 28 above)
Comment (skullbox305) [reply samples/sensor/fdc2x1x/README.rst:35]: Added your suggested note

File: samples/sensor/fdc2x1x/boards/nrf9160dk_nrf9160.overlay (hunk 29 above)
Comment (skullbox305) [reply samples/sensor/fdc2x1x/boards/nrf9160dk_nrf9160.overlay]: Made all channel bindings, except for `offset`, required

File: samples/sensor/fdc2x1x/src/main.c (hunk 31 above)
Comment (skullbox305) [reply samples/sensor/fdc2x1x/src/main.c:66]: I wanted to make the sample application work with both, the 2 channel and the 4 channel model of the sensor. I didn't know how else to get this information inside the sample. Should I instead make it only read 2 channels for both models?

**Comment by skullbox305** [comment]
Thank you for the feedback! I made the necessary changes and also left some comments. I also added the build_all test for the sensor. 
For some reason the compliance checks fail due to a trailing semicolon warning in the driver initialization macro. I don`t know how to fix this. Can I ignore this?

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 4 above)
Comment (mbolivar-nordic) [reply drivers/sensor/fdc2x1x/fdc2x1x.c:516]: > Nit: unnecessary whitespace.

Still here.

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 32]
```diff
@@ -0,0 +1,1002 @@
...
+	const struct fdc2x1x_config *cfg = dev->config;
+	uint8_t ch_supported;
+
+	if (cfg->fdc2x14) {
+		ch_supported = 4;
+	} else {
+		ch_supported = 2;
+	}
+
+	if (cfg->num_channels <= 0) {
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:882]: Taking another look at this line, it's using `<= 0` on an unsigned type. That's a bit misleading, please use `== 0`.

File: dts/bindings/sensor/ti,fdc2x1x.yaml (hunk 25 above)
Comment (mbolivar-nordic) [reply dts/bindings/sensor/ti,fdc2x1x.yaml:219]: > Same comments about default made above apply here also.

This `default` still does not have a justification in the description. The other ones look good now; thanks.

File: include/drivers/sensor.h (hunk 27 above)
Comment (mbolivar-nordic) [reply include/drivers/sensor.h:185]: This is a decision for the sensor maintainers; I can't comment on this one.

**Review by mbolivar-nordic — DISMISSED** [review]
> For some reason the compliance checks fail due to a trailing semicolon warning in the driver initialization macro. I don`t know how to fix this. Can I ignore it?

Thanks for your update! I have a couple of follow ups but this looks basically OK to me now.

You should be able to fix the compliance issue by rebasing onto the latest master. The checkpatch warnings were recently turned into hard CI errors, so we've been working through disabling the ones that should be turned off. Trailing semicolons were turned off in 967aaee36100cb07d77a089ebeddb06d36049c55 which is in master now, so a rebase will fix your PR's CI for that.

**Comment by mbolivar-nordic** [comment]
> One more thing: the driver must at least be build-tested by modifying tests/drivers/build_all.

Did you miss this one?

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 32 above)
Comment (skullbox305) [reply drivers/sensor/fdc2x1x/fdc2x1x.c]: Ok, now fixed

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 4 above)
Comment (skullbox305) [reply drivers/sensor/fdc2x1x/fdc2x1x.c]: sorry, my bad. Removed the whitespace now

File: dts/bindings/sensor/ti,fdc2x1x.yaml (hunk 25 above)
Comment (skullbox305) [reply dts/bindings/sensor/ti,fdc2x1x.yaml]: Sorry, forgot to remove the default

**Comment by skullbox305** [comment]
> 
> 
> > For some reason the compliance checks fail due to a trailing semicolon warning in the driver initialization macro. I don`t know how to fix this. Can I ignore it?
> 
> Thanks for your update! I have a couple of follow ups but this looks basically OK to me now.
> 
> You should be able to fix the compliance issue by rebasing onto the latest master. The checkpatch warnings were recently turned into hard CI errors, so we've been working through disabling the ones that should be turned off. Trailing semicolons were turned off in [967aaee](https://github.com/zephyrproject-rtos/zephyr/commit/967aaee36100cb07d77a089ebeddb06d36049c55) which is in master now, so a rebase will fix your PR's CI for that.

Thank you for the new feedback. I fixed all of your comments. Also the compliance check succeeded now. 
Looking forward to the feedback of the sensor maintainers for the other comments.

**Comment by skullbox305** [comment]
> 
> 
> > One more thing: the driver must at least be build-tested by modifying tests/drivers/build_all.
> 
> Did you miss this one?

No, I made the changes to `tests/drivers/build_all`. Did I forget to add something?

File: include/drivers/sensor.h (hunk 27 above)
Comment (MaureenHelm) [reply include/drivers/sensor.h:185]: I'm not a fan of adding channels without a type or standard unit. What does the quantity represent?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Overall really nice work, thank you for contributing!

File: include/drivers/sensor.h (hunk 27 above)
Comment (skullbox305) [reply include/drivers/sensor.h:185]: It can be represented as Capacitance or Frequency. I can add those conversion. Should I leave the Channel 0-3  numbering at the end of each type?

**Comment by mbolivar-nordic** [comment]
> No, I made the changes to `tests/drivers/build_all`. Did I forget to add something?

Sorry; I don't know how I missed that. Thanks.

File: samples/sensor/fdc2x1x/src/main.c (hunk 31 above)
Comment (mbolivar-nordic) [reply samples/sensor/fdc2x1x/src/main.c:66]: Let's keep it for now and revisit if we come up with a coherent policy in this regard.

File: include/drivers/sensor.h (hunk 27 above)
Comment (skullbox305) [reply include/drivers/sensor.h]: Added the conversions from raw to Capacitance and Freqzency.

**Comment by skullbox305** [comment]
> 
> 
> Overall really nice work, thank you for contributing!

Thank you!
I added the necessary conversion from raw to capacitance (pF) and frequency (MHz).

File: include/drivers/sensor.h [hunk 33]
```diff
@@ -174,6 +174,24 @@ enum sensor_channel {
...
+	/** CH0 Capacitance, in Picofarad **/
+	SENSOR_CHAN_CAPACITANCE_CH0,
+	/** CH1 Capacitance, in Picofarad **/
+	SENSOR_CHAN_CAPACITANCE_CH1,
+	/** CH2 Capacitance, in Picofarad **/
+	SENSOR_CHAN_CAPACITANCE_CH2,
+	/** CH3 Capacitance, in Picofarad **/
+	SENSOR_CHAN_CAPACITANCE_CH3,
+
+	/** CH0 Frequency, in MHz **/
```
Comment (pabigot) [inline include/drivers/sensor.h:186]: I'm not sure MHz is an ideal unit.  Hz would allow for frequencies like 0.5 Hz, i.e. every two seconds; or 0.000012 (a very inaccurate once-per-day).

This would limit the frequency to 2 GHz, though, which might be limiting in some applications.  Perhaps a future sensor API could select the best unit based on the value, and return either frequency in Hz or interval in some time unit, depending on which is more accurate.

This could be discussed in the API meeting.  Sticking with MHz or even kHz would be a reasonable resolution of such a discussion.

**Review by pabigot — COMMENTED** [review]
Some thoughts on sensor units.  Note that we're now in 2.5 stabilization so this won't get merged until the 2.6 merge window re-opens.

File: include/drivers/sensor.h (hunk 33 above)
Comment (skullbox305) [reply include/drivers/sensor.h:186]: I think you're right. Hz would make more sense. I'm not sure if >2GHz would be necessary for sensors, but who knows. Would it be possible to have both units, Hz now and MHz later if needed? This way you could use Hz for <2GHz and MHz for values above. 
Also I agree with your other suggestion. Having the flexibility to get the value returned in the most accurate unit would be really nice.

I will wait for the resolution of the API meeting though and then change it based on that. Could you give me an update on the resolution?

Comment (henrikbrixandersen) [reply include/drivers/sensor.h:186]: How about converting these sensor channel enums to be private to the FDC2X1X driver instead? Similar to how it is done for the MCUX ACMP sensor driver in https://github.com/zephyrproject-rtos/zephyr/blob/f71cd2aaf5e3f5942484a7ddb92e6192b2130dc6/include/drivers/sensor/mcux_acmp.h#L21?

Comment (carlescufi) [reply include/drivers/sensor.h:186]: API meeting 26th Jan 2021:

Moving the enums to the driver-specific header is the concluded best approach for this.

Comment (skullbox305) [reply include/drivers/sensor.h]: Thank you both for your suggestion!  I moved the channel enums to the driver-specific header now

File: drivers/sensor/fdc2x1x/fdc2x1x.c [hunk 34]
```diff
@@ -0,0 +1,1086 @@
...
+	static struct fdc2x1x_data fdc2x1x_data_##n = {			   \
+		.channel_buf = fdc2x1x_sample_buf_##n,			   \
+	};								   \
+									   \
+	const struct fdc2x1x_chx_config ch_cfg_##n[] = {		   \
+		DT_INST_FOREACH_CHILD(n, FDC2X1X_CH_CFG_INIT)		   \
+	};								   \
+									   \
+	static const struct fdc2x1x_config fdc2x1x_config_##n = {	   \
+		.i2c_name = DT_INST_BUS_LABEL(n),			   \
```
Comment (mbolivar-nordic) [inline drivers/sensor/fdc2x1x/fdc2x1x.c:1058]: Hi again. I just see one more round of churn before this PR LGTM from a driver model perspective.

We are making changes to the way that sensor devices access their buses for Zephyr 2.6. These mean that getting them by name using device_get_binding is discouraged.

You should now store the device directly in the configuration structure in ROM by moving these fields to `struct fdc2x1x_config` from `struct fdc2x1x_data`:

```
	const struct device *bus;
	const struct device *gpio;
```

These should now be initialized using `DEVICE_DT_GET`. For example, `.bus = DEVICE_DT_GET(DT_INST_BUS(n))`.

You should then check the device in the init routine using `device_is_ready`.

Finally, fields like `i2c_name` which exist only to store devicetree label properties should be removed. You can use `bus->name` instead if you need a name.

For an example, see the drivers/led/led_pwm.c changes in https://github.com/zephyrproject-rtos/zephyr/pull/32558.

Thanks for your patience!

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: drivers/sensor/fdc2x1x/fdc2x1x.c (hunk 34 above)
Comment (skullbox305) [reply drivers/sensor/fdc2x1x/fdc2x1x.c]: Hi. I addressed your comments. The bus and gpio devices are now initialized with `DEVICE_DT_GET` and checked  with `device_is_ready`. I also removed the `i2c_name` and both gpio labels.

**Review by MaureenHelm — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
Thanks! Sorry for the delay.

**Comment by carlescufi** [comment]
@skullbox305 could you please rebase?

**Comment by skullbox305** [comment]
@mbolivar-nordic Not a problem. Thank you and the others for the review! It was really helpful
@carlescufi I rebased it now
